File: Using_Java/LinkedLists/DoublyCircularLL.java
class NodeD
{
	public int 	iData;
	public NodeD nextPtr;
	public NodeD prevPtr;
	
	public NodeD()
	{
		iData 	= 0;
		nextPtr = null;
		prevPtr = null;
	}
	
	public NodeD(int iValue)
	{
		iData 	= iValue;
		nextPtr = null;
		prevPtr = null;
	}

}

interface LinkedList
{
	public void insertFirst(int iNo);
	public void insertLast(int iNo);
	
	public void deleteFirst();
	public void deleteLast();
	
	public void insertAtPos(int iNo,int iPos);
	public void deleteAtPos(int iPos);

}

class DoublyCircular implements LinkedList
{
	private int	iCount;
	private NodeD 	first;
	private NodeD 	last;
	
	public DoublyCircular()
	{
		iCount = 0;
		first  = null;
		last   = null;
	}
	
	public void displayNodes()
	{
		int iCnt = 0;
		
		NodeD tempPtr = first;
		
		for(iCnt = 0; iCnt < iCount; iCnt++)
		{
			System.out.print("|"+tempPtr.iData+"|<=>");
			tempPtr = tempPtr.nextPtr;
		}
		
	}
	
	public int countNodes()
	{
		return iCount;
	}

	public void insertFirst(int iNo)
	{
	
		NodeD newNode = null;
		
		newNode = new NodeD(iNo);
		
		if((first == null) && (last == null))
		{
			first = newNode;
			last  = newNode;
		}
		else
		{
			newNode.nextPtr = first;
			first.prevPtr = newNode;
			
			first = newNode;
		}
		
		last.nextPtr = first;
		first.prevPtr = last;
		
		iCount++;
	
	}
	
	public void insertLast(int iNo)
	{
		NodeD newNode = null;
		
		newNode = new NodeD(iNo);
		
		if((first == null) && (last == null))
		{
			first = newNode;
			last  = newNode;
		}
		else
		{
			last.nextPtr = newNode;
			newNode.prevPtr = last;
			
			last = newNode;
		}
		
		last.nextPtr = first;
		first.prevPtr = last;
		
		iCount++;
	
	
	}
	
	public void deleteFirst()
	{
		if((first == null) && (last == null))
		{
			System.out.println("\nLinked list is empty");
			return;
		}
		else if(first == last)
		{
			first = null;
			last  = null;
		}
		else
		{
			first = first.nextPtr;
			
			first.prevPtr = last;
			last.nextPtr  = first;
		}
		iCount--;	
	}
	
	public void deleteLast()
	{
		if((first == null) && (last == null))
		{
			System.out.println("\nLinked list is empty");
			return;
		}
		else if(first == last)
		{
			first = null;
			last  = null;
		}
		else
		{
			last = last.prevPtr;
			
			first.prevPtr = last;
			last.nextPtr  = first;
		}
		iCount--;
	}
	
	public void insertAtPos(int iNo,int iPos)
	{
		if((iPos < 0) || (iPos > iCount + 1))
		{
			System.out.println("Invalid position\n");
			return;
		}
		
		if(iPos == 1)
		{
			insertFirst(iNo);
		}
		else if(iPos == iCount + 1)
		{
			insertLast(iNo);
		}
		else
		{
			int iCnt = 0;
			NodeD tempPtr = first;
			
			for(iCnt = 1; iCnt < iPos - 1; iCnt++)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			NodeD newNode = null;			
			newNode = new NodeD(iNo);
			
			newNode.nextPtr = tempPtr.nextPtr;
			tempPtr.nextPtr = newNode;
			
			iCount++;
		}	
	}
	
	public void deleteAtPos(int iPos)
	{
		if((iPos < 0) || (iPos > iCount + 1))
		{
			System.out.println("Invalid position\n");
			return;
		}
		
		if(iPos == 1)
		{
			deleteFirst();
		}
		else if(iPos == iCount)
		{
			deleteLast();
		}
		else
		{
			int iCnt = 1;
			NodeD tempPtr  = first;
			NodeD tempPtrX = null; 
			
			for(iCnt = 1; iCnt < iPos - 1; iCnt++)
			{
				tempPtr = tempPtr.nextPtr;
			}
			
			tempPtrX = tempPtr.nextPtr;
			tempPtr.nextPtr = tempPtr.nextPtr.nextPtr;
			
			iCount--;
		}
	}
	

}

class DoublyCircularLL
{

	public static void main(java.lang.String Arg[])
	{
		DoublyCircular dObj = new DoublyCircular();
		
		dObj.insertFirst(101);
		dObj.insertFirst(51);
		dObj.insertFirst(21);
		dObj.insertFirst(11);

	System.out.println("============================================================================");
	
	
	System.out.println("\n----------------------------------------------------------------------------");
		System.out.println("Linked list after inserting nodes at first position : ");
		System.out.println("----------------------------------------------------------------------------");

		dObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+dObj.countNodes());
	
	System.out.println("\n----------------------------------------------------------------------------");	
		dObj.insertLast(111);
		dObj.insertLast(151);

		System.out.println("Linked list after inserting nodes at last position : ");
		System.out.println("----------------------------------------------------------------------------");		
		dObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		dObj.insertAtPos(55,4);

		System.out.println("Linked list after inserting node at given position : ");
		System.out.println("----------------------------------------------------------------------------");		
		dObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		dObj.deleteAtPos(4);

		System.out.println("Linked list after deleting node at given position : ");
		System.out.println("----------------------------------------------------------------------------");		
		dObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+dObj.countNodes());
	
	System.out.println("\n----------------------------------------------------------------------------");	
		
		dObj.deleteFirst();		
		
		System.out.println("Linked list after deleting node from first position : ");
		System.out.println("----------------------------------------------------------------------------");

		dObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n----------------------------------------------------------------------------");	
		
		dObj.deleteLast();
		System.out.println("Linked list after deleting node from last position : ");
		System.out.println("----------------------------------------------------------------------------");

		dObj.displayNodes();
		System.out.println("\nNumber of elements in linked lists are : "+dObj.countNodes());

	System.out.println("\n============================================================================");

	}

}

File: Using_Python/LinkedLists/DoublyCircularLL.py
from abc import ABC,abstractmethod

class NodeD:
	
	def __init__(self,iElement):
		
		self.iData = iElement
		self.nextPtr = None
		self.prevPtr = None
		
class LinkedLists(ABC):
	
	@abstractmethod
	def insertFirst(self,iNo):
		pass
			
	@abstractmethod
	def insertLast(self,iNo):
		pass
	
	@abstractmethod
	def deleteFirst(self):
		pass

	@abstractmethod		
	def deleteLast(self):
		pass
	
	@abstractmethod
	def insertAtPos(self,iNo,iPos):
		pass
	
	@abstractmethod	
	def deleteAtPos(self,iPos):
		pass


class DoublyCircularLL(LinkedLists):
	
	def __init__(self):
		
		self.iCount = 0
		self.first  = None
		self.last   = None
	
	def displayNodes(self):
		
		tempPtr = self.first
		
		print("<=>",end = ' ')
		for iCnt in range(0,self.iCount,1):
			
			print("|",tempPtr.iData,"|<=>",end = ' ')
			tempPtr = tempPtr.nextPtr
			
		print()
		
	def countNodes(self):
		
		return self.iCount	
	
	def insertFirst(self,iValue):
		
		newNode = None
		
		newNode = NodeD(iValue)
		
		if((self.first == None) and (self.last == None)):
			
			self.first = newNode
			self.last  = newNode
			
		else:
			
			newNode.nextPtr = self.first
			self.first = newNode
			
		self.first.prevPtr = self.last		
		self.last.nextPtr  = self.first
		
		self.iCount = self.iCount + 1
			
	def insertLast(self,iValue):
			
		newNode = None
		
		newNode = NodeD(iValue)
		
		if((self.first == None) and (self.last == None)):
			
			self.first = newNode
			self.last  = newNode
			
		else:
			self.last.nextPtr = newNode
			newNode.prevPtr = self.last
			
			self.last = newNode
			
		self.first.prevPtr = self.last		
		self.last.nextPtr  = self.first
		
		self.iCount = self.iCount + 1
			

		
	def deleteFirst(self):
		
		if(self.iCount == 0):
			
			print("Linked list is empty")
			return
		
		elif(self.iCount == 1):
			
			self.first = None
			self.last  = None
	
		else:
			self.first = self.first.nextPtr
			
			self.first.prevPtr = self.last		
			self.last.nextPtr  = self.first
			
		self.iCount = self.iCount - 1					
			
		
	def deleteLast(self):
		if(self.iCount == 0):
			
			print("Linked list is empty")
			return
		
		elif(self.iCount == 1):
			
			self.first = None
			self.last  = None
	
		else:
			self.last = self.last.prevPtr
			
			self.first.prevPtr = self.last		
			self.last.nextPtr  = self.first
			
		self.iCount = self.iCount - 1					
		
	def insertAtPos(self,iValue,iPos):
				
		if((iPos < 1) or (iPos > self.iCount + 1)):	
			
			print("Invalid position")
			
		if(iPos == 1):
			
			insertFirst(iValue)
			
		elif(iPos == self.iCount + 1):
			
			insertLast(iValue)
			
		else:
			
			tempPtr = self.first
			
			for iCnt in range(1,iPos - 1,1):
				
				tempPtr = tempPtr.nextPtr
			
			newNode = None
			
			newNode = NodeD(iValue)	
			
			newNode.nextPtr = tempPtr.nextPtr
			tempPtr.nextPtr = newNode
			
			self.iCount = self.iCount + 1

		
	def deleteAtPos(self,iPos):
		
		if((iPos < 1) or (iPos > self.iCount)):	
			
			print("Invalid position")
			
		if(iPos == 1):
			
			deleteFirst()
			
		elif(iPos == self.iCount):
			
			deleteLast()
			
		else:
			
			tempPtr = self.first
			tempPtrX = None
			
			for iCnt in range(1,iPos - 1,1):
				
				tempPtr = tempPtr.nextPtr
			
			tempPtrX = tempPtr.nextPtr
				
			tempPtr.nextPtr = tempPtr.nextPtr.nextPtr
			
			self.iCount = self.iCount - 1
			

def main():
	
	line = "-"*70
	
	print()
	print(line)
	print("Implementation of doubly circular linked list using python")
	print(line)

	dObj = DoublyCircularLL()		
	
	print("\nInserting elements at first position :")
	dObj.insertFirst(111)
	dObj.insertFirst(101)
	dObj.insertFirst(51)
	dObj.insertFirst(21)
	dObj.insertFirst(11)
	
	dObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(dObj.countNodes()))

	print("\nInserting elements at last position :")
	dObj.insertLast(151)
	dObj.insertLast(201)
	dObj.insertLast(211)

	dObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(dObj.countNodes()))
	
	print("\nInserting element at given position : ")
	dObj.insertAtPos(55,4)

	dObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(dObj.countNodes()))
	
	print("\nDeleting element from given position : ")
	dObj.deleteAtPos(4)

	dObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(dObj.countNodes()))
	
	print("\nDeleting element from first position : ")
	dObj.deleteFirst()

	dObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(dObj.countNodes()))
	
	print("\nDeleting element from last position : ")
	dObj.deleteLast()

	dObj.displayNodes()
	print("Number of nodes in linked lists are : {}".format(dObj.countNodes()))
	

if __name__ == "__main__":
	main()
